### ISSUE TITLE
Example SSLContext using TLS PSK KeyManager and TrustManager?

### ISSUE THREAD

**Issue by justincranford** [bcgit/bc-java#1884]
Does BC support an SSLContext provider that works with TLS PSK KeyManager/TrustManager?

I tried using SSLContext/BCJSSE with algorithms TLSv1.2 or TLSv1.3, but I get different errors depending what I try. I cannot find anything in the docs or examples to confirm if BC has a JSSE that supports TLS PSK, so I am asking to see if I should keep digging or not.

Summary of what I tried from [TlsPskUtil.java](https://github.com/justincranford/springs/blob/dev/springs-util-https/src/main/java/com/github/justincranford/springs/util/certs/util/TlsPskUtil.java) in my GitHub project:
```
public static SSLContext createPskSslContext(final SslBundle sslBundle) {
    final KeyManager[]   keyManagers   = new KeyManager[]   { createPskKeyManager(sslBundle) };
    final TrustManager[] trustManagers = new TrustManager[] { new PskTrustManager() };
    final SSLContext     sslContext    = SSLContext.getInstance("TLSv1.3", "BCJSSE");
    sslContext.init(keyManagers, trustManagers, SecureRandomUtil.SECURE_RANDOM);
}
```
```
public static SslContextFactory.Client createClientSslContextFactory(final SslBundle sslBundle) {
    final SSLContext sslContext = createPskSslContext(sslBundle);
    final SslContextFactory.Client sslContextFactory = new SslContextFactory.Client();
    sslContextFactory.setSslContext(sslContext);
    sslContextFactory.setIncludeProtocols(new String[] {"TLSv1.3"});
    sslContextFactory.setIncludeCipherSuites(new String[] {"TLS_AES_256_GCM_SHA384"});
    return sslContextFactory;
}
```
```
public static SslContextFactory.Server createServerSslContextFactory(final SslBundle sslBundle) {
    final SSLContext sslContext = createPskSslContext(sslBundle);
    final SslContextFactory.Server sslContextFactory = new SslContextFactory.Server();
    sslContextFactory.setSslContext(sslContext);
    sslContextFactory.setIncludeProtocols(new String[] {"TLSv1.3"});
    sslContextFactory.setIncludeCipherSuites(new String[] {"TLS_AES_256_GCM_SHA384"});
    return sslContextFactory;
}
```
- [createPskKeyManager(SslBundle)](https://github.com/justincranford/springs/blob/d4a64a829fbb91de5fa8c2217f9ae20e96060f4f/springs-util-https/src/main/java/com/github/justincranford/springs/util/certs/util/TlsPskUtil.java#L96C31-L96C50)
- [PskKeyManager class](https://github.com/justincranford/springs/blob/d4a64a829fbb91de5fa8c2217f9ae20e96060f4f/springs-util-https/src/main/java/com/github/justincranford/springs/util/certs/util/TlsPskUtil.java#L112-L145)
- [PskTrustManager class](https://github.com/justincranford/springs/blob/d4a64a829fbb91de5fa8c2217f9ae20e96060f4f/springs-util-https/src/main/java/com/github/justincranford/springs/util/certs/util/TlsPskUtil.java#L147-L164)

**Comment by justincranford** [comment]
TL;DR
I generate five bundles of TLS key material. They are injected at startup as text-based configuration properties, formatted as Spring Boot 3.x SSLBundle properties.
1. Client bundle for pskTLS
2. Server bundle for pskTLS
3. Client bundle for serverTLS
4. Client bundle for mutualTLS
5. Server bundle for serverTLS/mutualTLS (i.e. configured with server.ssl.clientAuth=want)

Running HttpsHelloWorldIT.java in my project triggers:
1. Reading the TLS key material,
2. Creating HTTP clients: serverTLS, mutualTLS, pskTLS
3. Creating HTTP server listeners: serverTLS/mutualTLS (localhost:8443), pskTLS (localhost:9443)
4. Running HTTP client-server tests: serverTLS (pass), mutualTLS (pass), pskTLS (fail)

**Comment by justincranford** [comment]
Details:

The pskTLS specific portion of the Spring Boot integration test for HTTP client->server testing.
- https://github.com/justincranford/springs/blob/d13b6aff6a8e3d2021b3a6a41f74c4d7172574fe/springs-util-https/src/test/java/com/github/justincranford/springs/util/certs/tls/HttpsHelloWorldIT.java#L95-L113

Spring Boot automatically loads the TLS key material for the cert-based listener port 8443.

Here is my custom util to auto-generate TLS key material. Text-based config properties are structured as Spring Boot 3.x SslBundle format.
- https://github.com/justincranford/springs/blob/d13b6aff6a8e3d2021b3a6a41f74c4d7172574fe/springs-util-https/src/main/java/com/github/justincranford/springs/util/certs/server/TlsInitializer.java#L176-L188

Here is my custom util to create PSK-based KeyManager+TrustManager for client, and server localhost:9443, from the SslBundle text-based properties.
- https://github.com/justincranford/springs/blob/d13b6aff6a8e3d2021b3a6a41f74c4d7172574fe/springs-util-https/src/main/java/com/github/justincranford/springs/util/certs/util/TlsPskUtil.java#L96-L164

Here is my custom util to create a PSK-based SSLContextFactory for client from the PSK-specific SslBundle text-based properties.
- https://github.com/justincranford/springs/blob/d13b6aff6a8e3d2021b3a6a41f74c4d7172574fe/springs-util-https/src/main/java/com/github/justincranford/springs/util/certs/util/TlsPskUtil.java#L66-L73

Here is my custom util to create a PSK-based SSLContextFactory for server localhost:9443, from the PSK-specific SslBundle text-based properties.
- https://github.com/justincranford/springs/blob/d13b6aff6a8e3d2021b3a6a41f74c4d7172574fe/springs-util-https/src/main/java/com/github/justincranford/springs/util/certs/util/TlsPskUtil.java#L75-L82


Spring Boot automatically creates a TLS Server localhost:8443 listener using certs. It handles incoming HTTP client connections uses serverTLS or mutualTLS, because ssl.client.auth=want.

Here is my custom code to create a TLS Server localhost:9443, for PSK-specific HTTP client connections.
- https://github.com/justincranford/springs/blob/d13b6aff6a8e3d2021b3a6a41f74c4d7172574fe/springs-util-https/src/main/java/com/github/justincranford/springs/util/certs/server/config/SpringsUtilHttpsServerPskConfiguration.java#L19-L34

**Comment by peterdettman** [comment]
Mentioned before in https://github.com/bcgit/bc-java/issues/1604 .

PSK is currently not supported in BCJSSE, mainly due to the lack of a standard (i.e. JSSE) API that would be needed to configure an SSLContext. We are not averse to including an appropriate extension API (apparently there are examples for other JSSE providers e.g. a PSKKeyManager for providing PSK credentials, distinct from the usual X509KeyManager).

**Comment by justincranford** [comment]
@winfriedgerlach yes this can be closed
